zebra-network/src/discoverer.rs

```diff
@@ -0,0 +1,166 @@
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::Stream;
+use pin_project::pin_project;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use tower::{
+    discover::{Change, Discover},
+    load::{CompleteOnResponse, PeakEwma, PeakEwmaDiscover},
+    BoxError,
+};
+
+use crate::{
+    constants,
+    peer::{Client, LoadTrackedClient},
+    protocol::external::types::Version,
+};
+
+type LoadTracker = PeakEwmaDiscover<UnboundedReceiverStream<DiscoveryEvent<Client>>>;
+type DiscoveryEvent<C> = Result<Change<SocketAddr, C>, BoxError>;
```

Comment:
Before Tests Nit: Is this generic ever used for multiple types?

```suggestion
type DiscoveryEvent = Result<Change<SocketAddr, Client>, BoxError>;
```